src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2113,31 +2227,46 @@ public static float Max(this IEnumerable<float> source)
         {
             if (source == null) throw Error.ArgumentNull("source");
             float value = 0;
-            bool hasValue = false;
-            foreach (float x in source)
+            
+            using (IEnumerator<float> e = source.GetEnumerator())
             {
-                if (hasValue)
-                {
-                    if (x > value || System.Double.IsNaN(value)) value = x;
-                }
-                else
+                if (!e.MoveNext()) throw Error.NoElements();
+                float x = e.Current;
+                if (Single.IsNaN(x)) return Single.NaN;
```

Comment:
Probably `using Single = System.Single;` at the top or replace the 15 occurrences of `Single.` with `System.Single.`